File: design-patterns/python/observer/fitness_data.py
from typing import List
from .fitness_data_subject import FitnessDataSubject
from .fitness_data_observer import FitnessDataObserver


class FitnessData(FitnessDataSubject):
    """Concrete Subject - maintains state and notifies observers"""
    
    def __init__(self):
        self._observers: List[FitnessDataObserver] = []
        self._steps = 0
        self._calories_burned = 0.0
        self._distance_covered = 0.0
    
    def register_observer(self, observer: FitnessDataObserver) -> None:
        self._observers.append(observer)
    
    def remove_observer(self, observer: FitnessDataObserver) -> None:
        if observer in self._observers:
            self._observers.remove(observer)
    
    def notify_observers(self) -> None:
        for observer in self._observers:
            observer.update(self)
    
    def set_fitness_data(self, steps: int, calories_burned: float, distance_covered: float) -> None:
        self._steps = steps
        self._calories_burned = calories_burned
        self._distance_covered = distance_covered
        print(f"FitnessData updated: Steps={steps}, Calories={calories_burned}, Distance={distance_covered}")
        self.notify_observers()
    
    # Getters
    @property
    def steps(self) -> int:
        return self._steps
    
    @property
    def calories_burned(self) -> float:
        return self._calories_burned
    
    @property
    def distance_covered(self) -> float:
        return self._distance_covered

File: design-patterns/golang/observer/fitness_data.go
package observer

// FitnessData is the concrete subject that manages observers and fitness data
type FitnessData struct {
	observers []FitnessDataObserver
	steps     int
	calories  int
	distance  float64
}

// NewFitnessData creates a new FitnessData instance
func NewFitnessData() *FitnessData {
	return &FitnessData{
		observers: make([]FitnessDataObserver, 0),
	}
}

func (f *FitnessData) RegisterObserver(observer FitnessDataObserver) {
	f.observers = append(f.observers, observer)
}

func (f *FitnessData) RemoveObserver(observer FitnessDataObserver) {
	for i, obs := range f.observers {
		if obs == observer {
			f.observers = append(f.observers[:i], f.observers[i+1:]...)
			break
		}
	}
}

func (f *FitnessData) NotifyObservers() {
	for _, observer := range f.observers {
		observer.Update(f.steps, f.calories, f.distance)
	}
}

// SetMeasurements updates the fitness data and notifies observers
func (f *FitnessData) SetMeasurements(steps int, calories int, distance float64) {
	f.steps = steps
	f.calories = calories
	f.distance = distance
	f.NotifyObservers()
}
